### DISCUSSION TITLE
Creating a tmux-like startup script

### DISCUSSION THREAD

**Discussion by danrot** [kovidgoyal/kitty#5908] [Q&A] [answered]
I want to switch over from kitty and avoid using tmux in the future. I've already set up the keybindings as I like them, but there is still one deal breaker for me: I've written a few tmux scripts, that allow me to set up a few panels and start commands in some of them. This looks something like this:

https://github.com/danrot/dotfiles/blob/4de91bb7ac0c09300f8661eb5da42c5bc25a03aa/tmux/script.tmux

This allows to me to execute something like the following and have my work environment set up:

```bash
script.tmux /some/directory/I/work/in
```

I've tried something similar with kitty, but failed so far. From what I see there are two potential solutions, but I did not get them to work for different reasons:

1.  In a [startup session](https://sw.kovidgoyal.net/kitty/overview/#startup-sessions) it does not seem possible to pass any parameters, as I do in my tmux script.
2. When trying to activate `allow_remote_control` in the config I think I should be able to implement similar scripts, but I could not get it to work...

The second option seems the one being closer. However, I was wondering if it is really necessary to activate remote control if I actually only want to execute a script from my local machine. I still tried it, and have written a script like the following:

```bash
#!/bin/bash

kitty @ set-tab-title $(basename $1)

kitty @ send-text "cd $1\n"

# kitty @ launch --cwd $1
```

If I use the script like that it changes the directory for the current kitty window, however, if I uncomment the last line the `cd` command is just output but not executed, which means I end up with two kitty windows, whereby both have the current working directoy not being set to the passed argument.

Does anybody have an idea why this is not working? I am trying this on a MacBook using the fish shell, in case that makes a difference.

**Comment by kovidgoyal** [comment]
You need "cd $1\r" the enter key sends \r not \n

Although for the use case you have actually described simply using kitty --directory whatever will work much better. tab titles will automatically be set to the directory by shell integration anyway.

**Reply by danrot** [reply]
I have already tried that before, the behavior is exactly the same. And using `--directory` is only a partial solution, since I also want to start `nvim` and other commands this way.

So the following script will output the directory listing:

```bash
#!/bin/bash

kitty @ set-tab-title $(basename $1)

kitty @ send-text "ls\r"
```
The result is a screen like this:
<img width="923" alt="image" src="https://user-images.githubusercontent.com/405874/213856207-d9ae2340-2601-45a2-aa75-c20219020783.png">

However, if add a second command after `send-text` it does not work anymore. So the following script:

```bash
#!/bin/bash

kitty @ set-tab-title $(basename $1)

kitty @ send-text "ls\r"

kitty @ launch
```

Will result in a screen like this:

<img width="868" alt="image" src="https://user-images.githubusercontent.com/405874/213856517-26d3fdfa-2b71-4bec-9b7e-f86cd591744a.png">

Mind that there is a `ls` output, which does not cause the directory listing to be output.

**Comment by kovidgoyal** [comment]
nvim can also output the working directory to the window title. But I am not following what you are trying to do. Are you running kitty and then running this script? Are you trying to run kitty from this script?

I *think* what you want to do is open kitty with some program running in it with a fixed title and working directory? If so write your script like this

```sh
#!/bin/zsh

exec kitty --session <(echo "tab_title $1\nlaunch --cwd=$1 $2")
```

And call it like this

```sh
myscript /path/to/working/dir program-to-execute
```

**Reply by danrot** [reply]
What you have suggested here opens a new OS windows, right? That's not really what I want... This is not really what I want... I am using the following script (which I have already posted as a link in the initial post):

```bash
#!/bin/bash

tmux rename-window $(basename $1)

tmux split-window -h
tmux split-window -v
tmux split-window -v

tmux send-keys -t 0 "cd $1" C-m
tmux send-keys -t 1 "cd $1" C-m
tmux send-keys -t 2 "cd $1" C-m
tmux send-keys -t 3 "cd $1" C-m

tmux resize-pane -t 0 -x 126
tmux select-pane -t 0

tmux send-keys "nvim ." C-m

tmux send-keys -t 2 "ls script.md | entr make script" C-m
tmux send-keys -t 3 "ls slides.md | entr make presentation" C-m
```

If I now execute this script in an already started tmux session, it will setup my environment to have nvim on the left, and on the right I have three panes, one with a shell, and two other that are already running a program. It does not open a new window.

See the following video to better understand what I expect the end result to be.

https://user-images.githubusercontent.com/405874/213866907-f54c76f1-ca44-4ba4-8905-95a7aec46d5d.mov

But maybe we can break that question down, there is only one problem, I think the rest I can work out on my own. If I activate `allow_remote_control` and I try to execute a script that uses `send-text` to execute something, then this won't work if another command comes afterwards. So the following script will "type" the `ls` command and execute it automatically, without me pressing enter after executing the script:

```bash
#!/bin/bash
kitty @ send-text "ls\n"
```

This works with both line endings (`\n` and `\r`). But if I add the script to also launch a new window afterwards, it does not work anymore. I am talking about a script like that:

```bash
#!/bin/bash
kitty @ send-text "ls\n"
kitty @ new-window
```

In that case the `ls` command will not be executed, no matter if I use `\n` or `\r` at the end, and that baffles me. If I know how I can make the latter script work I can figure out the rest on my own.

**Comment by kovidgoyal** [comment]
Works fine for me. most likely a timing issue between your shell
initializing/redrawing the prompt in response ot a resize event and the
text coming in. Insert a sleep or do the send text at the end after the
rest fo your commands.

**Reply by danrot** [reply]
Already tried that, the following script is also not working:

```bash
#!/bin/bash
kitty @ send-text "ls\n"
sleep 5
kitty @ new-window
```

The result looks like that:


https://user-images.githubusercontent.com/405874/213872211-d63e2f4c-b419-4de3-bcd2-34e1e6d272e4.mov

**Comment by kovidgoyal** [comment, marked as answer]
Wait are you running this script at a prompt? If so the text will show up before the next prompt is drawn, since there is a subsequent command to run. You need to have your send text as the last command.

**Reply by danrot** [reply]
Thank you, that worked! It is a bit confusing though 🙈 I am posting the final solution here, for persons finding that discussion:

```bash
#!/bin/bash
kitty @ set-window-title $(basename $1)
kitty @ launch --cwd $1
kitty @ launch --cwd $1
kitty @ launch --cwd $1
kitty @ focus-window --match "recent:3"

kitty @ send-text --match "recent:1" "ls script.md | entr make script\n"
kitty @ send-text --match "recent:2" "ls slides.md | entr make presentation\n"
kitty @ send-text --match "recent:0" "cd $1;nvim .\n"
```

This does more or less the same as I've done before with tmux.
